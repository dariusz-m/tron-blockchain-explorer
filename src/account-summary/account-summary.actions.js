export const AccountSummaryActionTypes = {
    ACCOUNT_DETAILS_HAVE_BEEN_LOADED: "ACCOUNT_DETAILS_HAVE_BEEN_LOADED",
    ACCOUNT_DETAILS_HAVE_NOT_BEEN_LOADED: "ACCOUNT_DETAILS_HAVE_NOT_BEEN_LOADED",
    LOAD_ACCOUNT_DETAILS: "LOAD_ACCOUNT_DETAILS",
    LOAD_TRANSACTIONS_BY_ACCOUNT: "LOAD_TRANSACTIONS_BY_ACCOUNT",
    TRANSACTION_BY_ACCOUNT_HAVE_BEEN_LOADED: "TRANSACTION_BY_ACCOUNT_HAVE_BEEN_LOADED",
    TRANSACTION_BY_ACCOUNT_HAVE_NOT_BEEN_LOADED: "TRANSACTION_BY_ACCOUNT_HAVE_NOT_BEEN_LOADED",
};


export const loadAccountDetails = (address) => {
    return {
        type: AccountSummaryActionTypes.LOAD_ACCOUNT_DETAILS,
        payload: address
    };
};

export const accountDetailsHaveBeenLoaded = (accountDetails) => {
    return {
        type: AccountSummaryActionTypes.ACCOUNT_DETAILS_HAVE_BEEN_LOADED,
        payload: accountDetails
    };
};

export const accountDetailsHaveNotBeenLoaded = () => {
    return {
        type: AccountSummaryActionTypes.ACCOUNT_DETAILS_HAVE_NOT_BEEN_LOADED,
    };
};

export const loadTransactionsByAccount = (address) => {
    return {
        type: AccountSummaryActionTypes.LOAD_TRANSACTIONS_BY_ACCOUNT,
        payload: address
    };
};

export const transactionsByAccountHaveBeenLoaded = (transactions) => {
    return {
        type: AccountSummaryActionTypes.TRANSACTION_BY_ACCOUNT_HAVE_BEEN_LOADED,
        payload: transactions
    };
};

export const transactionsByAccountHaveNotBeenLoaded = () => {
    return {
        type: AccountSummaryActionTypes.TRANSACTION_BY_ACCOUNT_HAVE_NOT_BEEN_LOADED,
    };
};
