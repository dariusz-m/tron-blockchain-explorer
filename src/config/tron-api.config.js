export const TRON_ORGIN = "https://api.tronscan.org";
