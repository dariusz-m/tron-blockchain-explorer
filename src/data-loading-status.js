export const DATA_LOADING_STATUS = {
    DATA_HAVE_BEEN_LOADED: "DATA_HAVE_BEEN_LOADED",
    DATA_HAVE_NOT_BEEN_LOADED: "DATA_HAVE_NOT_BEEN_LOADED",
    DATA_PROCESSING: "DATA_PROCESSING",
};
