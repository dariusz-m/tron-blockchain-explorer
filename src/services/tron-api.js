import HttpClient from "@tronprotocol/wallet-api/src/client/http";

export const TronBlockchainClient = new HttpClient();
