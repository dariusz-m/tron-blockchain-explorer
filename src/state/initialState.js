const initialState = {
    router: {"location": null},
    tokens: [],
    selectedBlock: {number: -1, witnessId: -1, hash: "", parentHash: ""}
};

export default initialState;
