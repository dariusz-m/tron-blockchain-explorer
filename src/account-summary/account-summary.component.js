import React from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {bindActionCreators} from "redux";
import ReactList from 'react-list';

import {loadAccountDetails, loadTransactionsByAccount} from './account-summary.actions';
import {DATA_LOADING_STATUS} from "../data-loading-status";

export class AccountSummary extends React.Component {
    constructor() {
        super();
        this.renderTransaction = this.renderTransaction.bind(this);
    }

    componentDidMount() {
        if(this.props.statusOfDataLoading === DATA_LOADING_STATUS.DATA_HAVE_NOT_BEEN_LOADED) {
            this.props.actions.loadAccountDetails(this.props.match.params.address);
        }
        this.props.actions.loadTransactions(this.props.match.params.address);
    }

    renderTransaction(index, key) {
        return (
            <div key={key}>
                <p>{this.props.transactions[index].amount}</p>
                <p>{this.props.transactions[index].block}</p>
                <p>{this.props.transactions[index].hash}</p>
                <p>{this.props.transactions[index].timestamp}</p>
                <p>{this.props.transactions[index].tokenName}</p>
                <p>{this.props.transactions[index].transferFromAddress}</p>
                <p>{this.props.transactions[index].transferToAddress}</p>
            </div>
        );
    }

    render() {
        return (
            <div>
                <h1>Account</h1>
                <p>{this.props.account.address}</p>
                <p>{this.props.account.balance}</p>
                <p>{this.props.account.name}</p>
                <div>

                    {Object.keys(this.props.account.tokenBalances).map((tokenName) => {
                        return <p key={tokenName}>{tokenName}: {this.props.account.tokenBalances[tokenName]}</p>;
                    })}
                </div>
                <h2>Transactions</h2>
                <div style={{overflow: 'auto', maxHeight: 200}}>
                    <ReactList
                        itemRenderer={this.renderTransaction}
                        length={this.props.transactions.length}
                    />
                </div>
            </div>
        );
    }
}

AccountSummary.propTypes = {
    account: PropTypes.shape({
        address: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        /* e.g {"Shasta": 1024, "By A Nose": 5300, "Head Coin": 1301}*/
        tokenBalances: PropTypes.object.isRequired,
    }).isRequired,
    transactions: PropTypes.arrayOf(PropTypes.shape({
        amount: PropTypes.number.isRequired,
        block: PropTypes.number.isRequired,
        hash: PropTypes.string.isRequired,
        timestamp: PropTypes.number.isRequired,
        tokenName: PropTypes.string.isRequired,
        transferFromAddress: PropTypes.string.isRequired,
        transferToAddress: PropTypes.string.isRequired
    })).isRequired,
    statusOfDataLoading: PropTypes.string.isRequired,
    actions: PropTypes.shape({
        loadTransactions: PropTypes.func.isRequired,
        loadAccountDetails: PropTypes.func.isRequired,
    }),
};

function mapStateToProps(state) {
    return {
        account: state.accountSummary.account,
        transactions: state.accountSummary.transactions,
        statusOfDataLoading: state.accountSummary.statusOfDataLoading,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            loadTransactions: loadTransactionsByAccount,
            loadAccountDetails: loadAccountDetails,
        }, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(AccountSummary);
