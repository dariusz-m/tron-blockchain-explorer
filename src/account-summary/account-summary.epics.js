import { from, of } from 'rxjs';
import { mergeMap, map, catchError, pluck} from 'rxjs/operators';
import { ajax } from 'rxjs/ajax';

import {
    AccountSummaryActionTypes,
    transactionsByAccountHaveBeenLoaded,
    transactionsByAccountHaveNotBeenLoaded,
    accountDetailsHaveBeenLoaded,
    accountDetailsHaveNotBeenLoaded
} from './account-summary.actions';
import {TRON_ORGIN} from "../config/tron-api.config";

export const loadTransactionsByAccount = action$ =>
    action$.ofType(AccountSummaryActionTypes.LOAD_TRANSACTIONS_BY_ACCOUNT)
        .pipe(
            mergeMap((action) =>
                from(
                    ajax(`${TRON_ORGIN}/api/transaction?sort=-timestamp&address=${action.payload}`)
                ).pipe(
                    pluck('response','data'),
                    map((transactions) => transactionsByAccountHaveBeenLoaded(transactions)),
                    catchError((error) => {
                        console.error(error);
                        return of(transactionsByAccountHaveNotBeenLoaded());
                    })
                )
            ),
        );

export const loadAccountDetails = action$ =>
    action$.ofType(AccountSummaryActionTypes.LOAD_ACCOUNT_DETAILS)
        .pipe(
            mergeMap((action) =>
                from(
                    ajax(`${TRON_ORGIN}/api/account?sort=-balance&limit=1&address=${action.payload}`)
                ).pipe(
                    pluck('response','data'),
                    map((accountDetails) => accountDetailsHaveBeenLoaded(accountDetails[0])),
                    catchError((error) => {
                        console.error(error);
                        return of(accountDetailsHaveNotBeenLoaded());
                    })
                )
            ),
        );

