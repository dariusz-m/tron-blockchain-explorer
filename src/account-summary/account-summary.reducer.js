import initialState from '../state/initialState';
import {AccountSummaryActionTypes} from './account-summary.actions';
import {DATA_LOADING_STATUS} from "../data-loading-status";

export default (state = initialState.accountSummary, action) => {
    switch (action.type) {
        case AccountSummaryActionTypes.ACCOUNT_DETAILS_HAVE_BEEN_LOADED:
            return {...state, account: action.payload, statusOfDataLoading: DATA_LOADING_STATUS.DATA_HAVE_BEEN_LOADED};
        case AccountSummaryActionTypes.LOAD_ACCOUNT_DETAILS:
            return {...state, statusOfDataLoading: DATA_LOADING_STATUS.DATA_PROCESSING};
        case AccountSummaryActionTypes.TRANSACTION_BY_ACCOUNT_HAVE_BEEN_LOADED:
            return {...state, transactions: action.payload};
        default:
            return state;
    }
};
