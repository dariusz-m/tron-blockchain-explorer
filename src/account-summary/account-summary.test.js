import React from 'react';
import renderer from 'react-test-renderer';
import nock from 'nock';

import configureStore from "../state";
import initialState from "../state/initialState";
import {DATA_LOADING_STATUS} from "../data-loading-status";
import {AccountSummary} from "./account-summary.component";
import {loadAccountDetails, loadTransactionsByAccount} from "./account-summary.actions";
import {TRON_ORGIN} from "../config/tron-api.config";

jest.mock('react-list', () => 'div');


describe('Account summary tests', () => {
    let store;
    let accountDetails = {};
    let transactions = [];

    beforeEach(() => {
        store = configureStore();
        accountDetails = {
            address: "27d3byPxZXKQWfXX7sJvemJJuv5M65F3vjS",
            balance: 41935732000000000,
            name: "Devaccount",
            tokenBalances: {Evan: 88999}
        };
        transactions = [
            {
                hash: "532c288aab0b5a8355e5baaef8c77891b4a9ef1feaf44202ce2ddfff54903930",
                block: 0,
                timestamp: 0,
                transferFromAddress: "7YxAaK71utTpYJ8u4Zna7muWxd1pQwimpGxy8",
                transferToAddress: "27d3byPxZXKQWfXX7sJvemJJuv5M65F3vjS",
                amount: 75000000000000000,
                tokenName: "TRX",
            },
            {
                hash: "5d5bbc7a256058e15887342fba8388cfdbc7d00dbde800e5a29204efe2b15a13",
                block: 0,
                timestamp: 0,
                transferFromAddress: "7YxAaK71utTpYJ8u4Zna7muWxd1pQwimpGxy8",
                transferToAddress: "27WtBq2KoSy5v8VnVZBZHHJcDuWNiSgjbE3",
                amount: 9223372036854776000,
                tokenName: "TRX",
            },
        ];
    });

    it('Renders correctly', () => {
        const actions = {
            loadTransactions: jest.fn(),
            loadAccountDetails: jest.fn(),
        };
        const match = {params: {address: 1}};

        const tree = renderer.create(
            <AccountSummary
                actions={actions}
                transactions={transactions}
                account={accountDetails}
                statusOfDataLoading={DATA_LOADING_STATUS.DATA_HAVE_BEEN_LOADED}
                match={match}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Account details has been loaded', (done) => {
        expect.hasAssertions();
        nock(TRON_ORGIN)
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .get(`/api/account?sort=-balance&limit=1&address=${accountDetails.address}`)
            .reply(200, {data: [accountDetails]});
        const expectedState = {
            ...initialState,
            accountSummary: {
                ...initialState.accountSummary,
                statusOfDataLoading: DATA_LOADING_STATUS.DATA_HAVE_BEEN_LOADED,
                account: accountDetails
            }
        };

        store.dispatch(loadAccountDetails(accountDetails.address));

        store.subscribe(() => {
            expect(store.getState()).toEqual(expectedState);
            done();
        });
    });

    it('Transactions for account have been loaded', (done) => {
        expect.hasAssertions();
        nock(TRON_ORGIN)
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .get(`/api/transaction?sort=-timestamp&address=${accountDetails.address}`)
            .reply(200, {data: [...transactions]});
        const expectedState = {
            ...initialState,
            accountSummary: {
                ...initialState.accountSummary,
                transactions: [...transactions]
            }
        };

        store.dispatch(loadTransactionsByAccount(accountDetails.address));

        store.subscribe(() => {
            expect(store.getState()).toEqual(expectedState);
            done();
        });
    });

});

