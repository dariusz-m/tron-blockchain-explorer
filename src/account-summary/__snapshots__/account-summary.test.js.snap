// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account summary tests Renders correctly 1`] = `
<div>
  <h1>
    Account
  </h1>
  <p>
    27d3byPxZXKQWfXX7sJvemJJuv5M65F3vjS
  </p>
  <p>
    41935732000000000
  </p>
  <p>
    Devaccount
  </p>
  <div>
    <p>
      Evan
      : 
      88999
    </p>
  </div>
  <h2>
    Transactions
  </h2>
  <div
    style={
      Object {
        "maxHeight": 200,
        "overflow": "auto",
      }
    }
  >
    <div
      itemRenderer={[Function]}
      length={2}
    />
  </div>
</div>
`;
